`PyArrayDescr` methods like in `np.dtype` + some missing stuff in `npyffi`

aldanor opened the pull request:
- The `npyffi` stuff has been pulled out of [`pyo3-type-desc`](https://github.com/aldanor/pyo3-type-desc).
- Methods in `PyArrayDescr` mostly replicate `np.dtype` object, plus a few methods replicating C macros (this will greatly simplify writing tests in `pyo3-type-desc`, plus makes `PyArrayDescr` object somewhat useful).

aldanor commented:
There's still a few TODOs and questions around that will need to be resolved, plus I wanted to add at least some trivial tests for each of the methods, and add a `new()` that replicates `np.dtype()` constructor.

aldanor commented on src/npyffi/flags.rs:
```diff
@@ -62,3 +62,21 @@ pub const NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE: npy_uint32 = 0x40000000;
+pub const NPY_NEEDS_PYAPI: npy_char = 0x10;
+pub const NPY_USE_GETITEM: npy_char = 0x20;
+pub const NPY_USE_SETITEM: npy_char = 0x40;
+pub const NPY_ALIGNED_STRUCT: npy_char = -128; // 0x80
```
These flags are used in `c_char` context (`dtype->flags`), hence this weirdness

adamreichold commented on src/npyffi/array.rs:
```diff
@@ -394,6 +394,12 @@ pub unsafe fn PyArray_CheckExact(op: *mut PyObject) -> c_int {
 }
 
+// these are under `#if NPY_USE_PYMEM == 1` which seems to be always defined as 1
+pub use pyo3::ffi::{
```
I don't really understand why we need to re-export those in the first place? Why not use them directly from `pyo3`?

adamreichold replied:
(Concerning the comment, I think these would only be enabled if CPython is built for usage with e.g. Valgrind, c.f. https://github.com/python/cpython/blob/main/Misc/README.valgrind)

davidhewitt replied:
I think it makes sense to re-export with new names, to match the numpy header.

aldanor commented on src/npyffi/types.rs:
```diff
@@ -201,3 +201,70 @@ pub struct npy_stride_sort_item {
+#[derive(Debug, Clone, Copy)]
+// NPY_TYPEKINDCHAR doesn't exist in the header, but these enum values are not
+// related to NPY_TYPECHAR although being stuffed into it (type kinds, not type codes)
+pub enum NPY_TYPEKINDCHAR {
```
In the original header these are part of `NPY_TYPECHAR`, but they have repeated values + a comment that they don't belong there.

davidhewitt replied:
Is there a strong reason why we should differ from the C header?

aldanor replied:
Just one reason really - Rust enums can't have repeated values... (all 5 of these values already occur in `NPY_TYPECHAR')

aldanor replied:
(perhaps the comment should reflect that as well)

aldanor replied:
(Updated the comment preceding this enum, hope it's more clear now)

adamreichold commented on src/dtype.rs:
```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+            return None;
+        }
+        let names = unsafe { PyTuple::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).names) };
+        <_>::extract(names).ok()
```
This (and below) feels a bit too clever. Couldn't you just call out the trait by name, i.e. `FromPyObject::extract`?

aldanor replied:
Fair enough, fixed (I guess I love turbofish too much since this is the first thing that comes to mind).

adamreichold commented on src/dtype.rs:
```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+            }
+            Ok(fields)
+        })()
+        .ok()
```
Why swallow up the `Result` into `Option` instead of returning `PyResult`?

aldanor replied:
Because `Option` means there's no fields. So it would have to be `Option<PyResult<...>>`.

Thing is, this `PyResult<>` will be always ok, those fields are not mutable from outside, are created in the C code and are relied upon in the entire numpy.

adamreichold replied:
Then `unwrapping` seems appropriate. If not, `PyResult<Option<_>, _>` would be the more fitting signature IMHO. (Similary for the other methods which turn `PyResult` into `Option`.)

aldanor replied:
I think unwrapping is ok here (and if anyone can make it panic from within Python, it should be a bug submitted to numpy upstream).

In which case, is an iterator or a collected (btreemap) preferable?

adamreichold replied:
Let's resolve this discussion in favour of https://github.com/PyO3/rust-numpy/pull/261/files#r788000640 to reduce the clutter.

aldanor commented on src/dtype.rs:
```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+    /// Note: titles (the optional 3rd tuple element) are ignored.
+    ///
+    /// Equivalent to [`np.dtype.fields`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.fields.html).
+    pub fn fields(&self) -> Option<BTreeMap<&str, (&PyArrayDescr, usize)>> {
```
Questions here:

- Should it be an `Option<BTreeMap>`?
- Alternatively, it could be an `Option<impl Iterator<Item = (&str, (&PyArrayDescr, usize))> + '_>`, any better?
- NumPy sort-of-guarantees that those fields will be correct tuples of correct size, maybe it's ok to just `.unwrap()`?

adamreichold replied:
I think the iterator would either require unwrapping or having `Result` as the item type. It feels overcomplicated to me. Personally, I am admittedly not to happy about the additional API surface in the first, especially a lot of this duplicates macros which would change theoretically upstream without us noticing.

adamreichold replied:
If we do the unpacking, then `Option<impl Iterator>` seems preferable as it makes fewer decisions for calling code (which for example might not care about ordering).

Returning `impl Trait` in a public API might be considered bad style though is it very difficult to store such a type anywhere (as it cannot be named), but I think this not the most pressing concern here as the iterator has a limited lifetime derived from `&self`.

But what I actually wonder is whether we shouldn't just return `&PyDict` instead of doing the unpacking ourselves. Do I understand it correctly that this is to simplify test code where working directly with `PyDict` and `PyTuple` etc. is rather unergonomic?

aldanor replied:
>  additional API surface in the first

Do you mean the behaviour and contents of `->fields`? 

That exact getter wasn't changed in the last 16 years at least since that's where the explicit getter was introduced and the immutable dictproxy was added ([link](https://github.com/numpy/numpy/commit/95fc4486e429c82a872f3debcf09b0273fe2523e)), and it existed in the same form, but less formally, and before that. Given the amount of external stuff relying on numpy, it's pretty much set in stone by now.

If you mean the 'additional API surface' in general - almost all of these getters simply expose various fields of `PyArray_Descr` without any alterations, providing the same API (or close to that) as the numpy itself. I'm not sure I understand.

adamreichold replied:
> That exact getter wasn't changed in the last 16 years at least since that's where the explicit getter was introduced and the immutable dictproxy was added (link), and it existed in the same form, but less formally, and before that. Given the amount of external stuff relying on numpy, it's pretty much set in stone by now.

Tried to capture this by "theoretically" as C code using the macros which is not rebuilt would have the same issue with such changes.

> If you mean the 'additional API surface' in general - almost all of these getters simply expose various fields of PyArray_Descr without any alterations, providing the same API (or close to that) as the numpy itself. I'm not sure I understand.

They still raise design questions like the `impl Iterator` or `BTreeMap` choice. I also fear that this API will see little usage outside of special circumstances where the FFI API might provide a reasonable alternative, i.e. probably not when processing arrays in Python extensions.

aldanor replied:
@adamreichold @davidhewitt Would everyone be happy with the suggestion in https://github.com/PyO3/rust-numpy/pull/261#issuecomment-1016958418? If yes, I'll replace `fields` with `get_field` which will keep the API leaner and proceed on with writing tests for all of this. This will be the only `get_`-prefixed method then, which should make it easier to notice it does something less trivial than the other ones.

adamreichold replied:
Yes, I like the solution of exposing `get_field` instead of `fields`.

aldanor commented on src/dtype.rs:
```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+        if !self.has_fields() {
+            return None;
+        }
+        let names = unsafe { PyTuple::from_borrowed_ptr(self.py(), (*self.as_dtype_ptr()).names) };
```
Maybe my lack of `pyo3` knowledge, looks like an overkill to incref here for just one line of code? Is it possible to do it without `from_borrowed_ptr`, or is this the intended way to do it?

adamreichold replied:
Are you sure there is reference increment involved here, c.f. https://docs.rs/pyo3/0.15.1/src/pyo3/conversion.rs.html#543-548? I think that happens only when creating e.g. `Py<T>`?

aldanor replied:
Ohh, fair enough - yea, I was looking at source code for the wrong impl. It's all ok then.

aldanor commented on src/dtype.rs:
```diff
@@ -94,12 +106,202 @@ impl PyArrayDescr {
+    /// These are roughly ordered from least-to-most precision.
+    ///
+    /// Equivalent to [`np.dtype.num`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.num.html).
+    pub fn num(&self) -> c_int {
```
This has been renamed from `get_typenum()` to stay close to `np.dtype` Python API. Very unlikely it was used anywhere (and in the internal code there was only one reference).

adamreichold replied:
Personally, I don't mind the rename seeing our version number and considering the theme of deriving the names from `numpy.dtype`.

adamreichold commented:
> The NPY_TYPES enum is marked as u32, however in C code it's always a c_int

I don't think that it is possible to reference the type alias in a `repr` directive which is probably why it has this representation in the first place.

> so it would have at least some use in a public API of PyArrayDescr, but then it would need a separate type.

Personally, I would actually prefer it not becoming part of the API again if possible. Put differently, I see API surface as a liability that should be avoided if possible.

aldanor commented:
@adamreichold Fixed all todos and a few things noted above, let me know if you think anything else is needed to change/add/remove.

(I still need to write tests for all of this; although it's all quite trivial it should all get triggered at least once somewhere in the test suite.)

As for `fields` - I've changed them to an iterator of unwrapped items in this commit - BUT, here's another idea: what if we change it to a simple
```rust
pub fn get_field(&self, name: &str) -> PyResult<(&PyArrayDescr, usize)>
```
And then we don't have maps, iterators and excessive unwrapping, it all lands onto the user and we keep it reasonably simple. Here, `get_` prefix indicates that some non-trivial action will be performed and in this case it's also fallible (btw, to be consistent with other names here, I'd also suggest to rename `get_type()` to `typeobj` or something, since it doesn't do anything and just returns `self->typeobj`).

Given that there's already `fn names(&self) -> Option<Vec<&str>>`, this should be more than sufficient.

davidhewitt commented on src/dtype.rs:
````diff
@@ -57,21 +68,6 @@ impl PyArrayDescr {
-    ///    assert_eq!(dtype.get_type().name().unwrap().to_string(), "float64");
-    /// });
-    /// ```
-    pub fn get_type(&self) -> &PyType {
````
Should we consider marking as deprecated instead of removing immediately, to make it easier for users to migrate? It may not be immediately obvious what they should migrate to.

aldanor replied:
That crossed my mind as well - sure, we can leave is as deprecated for a version or so.

davidhewitt replied:
👍 thanks, can even suggest what to migrate to in the deprecation note!

davidhewitt commented on src/dtype.rs:
```diff
@@ -94,12 +90,207 @@ impl PyArrayDescr {
-    /// Retrieves the
-    /// [enumerated type](https://numpy.org/doc/stable/reference/c-api/dtype.html#enumerated-types)
-    /// for this type descriptor.
-    pub fn get_typenum(&self) -> c_int {
```
Similar q about deprecation here.

aldanor replied:
This one ^ I don't think we need to deprecate because it doesn't exist in the latest stable (0.15.1).

davidhewitt reviewed (commented):
👍 just skim read, got a few quick thoughts. Thanks!

Would it be worth adding some tests for the new dtype methods?

davidhewitt commented:
> (I still need to write tests for all of this; although it's all quite trivial it should all get triggered at least once somewhere in the test suite.)

If there's interest, you might want to add a coverage job to CI. You can steal configuration from pyo3 :)

aldanor commented:
> If there's interest, you might want to add a coverage job to CI. You can steal configuration from pyo3 :)

Looks pretty nice, good point! (never had a chance to use ci coverage with rust in particular for some reason, but from a brief look at pyo3's codecov reports looks pretty cool). I'll try to open a separate pr for that tomorrow here.

aldanor commented:
Ok:

- Added tests for all of the methods (on a scalar, subarray and a struct type)
- Removed `subdtype` and `is_unsized`, they are redundant
- Replaced `fields` with `get_field`
- Fixed `base` and its return type (no more `Option`)
- Added `PyArrayDescr:::get_type()` back as hidden/deprecated
- (various cleanup etc)

aldanor commented:
Ok, clippy's finally happy. Added a changelog entry as well. So unless there's further comments this should be good to go.

adamreichold reviewed (approved):
> Ok, clippy's finally happy. Added a changelog entry as well. So unless there's further comments this should be good to go.

I think so too. Will leave it open until tomorrow if anybody else wants to have a look.

(It is a matter of personal preferences, but I do find the changelog entry a bit on the wordy end of the scale. This could undermine the changelog's purpose of quickly getting an overview of the changes. The argument does not seem strong enough to me to ask for changes, I just wanted to mention it for consideration.)

kngwyu commented on src/dtype.rs:
```diff
@@ -94,12 +91,202 @@ impl PyArrayDescr {
+    /// The names are ordered according to increasing byte offset.
+    ///
+    /// Equivalent to [`np.dtype.names`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.names.html).
+    pub fn names(&self) -> Option<Vec<&str>> {
```
Same as above

adamreichold replied:
https://github.com/PyO3/rust-numpy/pull/261#discussion_r790141617

kngwyu commented on src/npyffi/types.rs:
```diff
@@ -201,3 +201,72 @@ pub struct npy_stride_sort_item {
+    NPY_IGNORE = b'|',
+}
+
+impl NPY_BYTEORDER_CHAR {
```
I like this approach 👍🏼

kngwyu reviewed (approved):
Well done, thank you!

adamreichold commented on src/dtype.rs:
```diff
@@ -94,12 +91,202 @@ impl PyArrayDescr {
+    /// Returns shape tuple of the sub-array if this dtype is a sub-array, and `None` otherwise.
+    ///
+    /// Equivalent to [`np.dtype.shape`](https://numpy.org/doc/stable/reference/generated/numpy.dtype.shape.html).
+    pub fn shape(&self) -> Option<Vec<usize>> {
```
But isn't an empty `Vec` a separate yet valid case, i.e. a zero-dimensional sub-array, or is this not possible?

kngwyu replied:
> or is this not possible?

I don't know actually.... this is what I meant by
> if the returned Vec here is always non-empty?

aldanor replied:
Ok, I've checked in Python, and from the looks of it:

- You can create an empty struct dtype, which is a valid case, so `names()` has to remain behind an `Option`:
  ```python
  >>> np.dtype([]).fields
  mappingproxy({})
  >>> np.dtype('f8').fields
  None
  ```
- With subarrays though... I'm not sure you can *create* a zero-dimensional subarray dtype through normal constructors, as it gets flattened down to the underlying scalar; also, you can request the shape of normal scalar types and it returns empty tuple:
  ```python
  >>> np.dtype(('f8', ()))
  np.dtype(('f8', ()))
  >>> np.dtype('f8').shape
  ()
  ```
  So if we want to be closer to `np.dtype`'s API, we can remove `Option<>` from `shape()` and return an empty vec for non-subarray types?

adamreichold replied:
Mostly due to

> also, you can request the shape of normal scalar types and it returns empty tuple: 

I think leaving away the `Option` layer would indeed be better w.r.t. consistency.

aldanor replied:
Btw I made an error while copying the snippet above, the correct version is:
```python
>>> np.dtype(('f8', ()))
dtype('float64')
```
^ so here you can see that it doesn't let you create a 0-dim subarray type and flattens it down.

Anyway, I've removed the option layer, so you can now always request the shape (same as in the numpy itself).

aldanor commented:
> (It is a matter of personal preferences, but I do find the changelog entry a bit on the wordy end of the scale. This could undermine the changelog's purpose of quickly getting an overview of the changes. The argument does not seem strong enough to me to ask for changes, I just wanted to mention it for consideration.)

@adamreichold I've noticed that as well. I guess I'm just used to more lengthy changelogs like the [one](https://github.com/aldanor/hdf5-rust/blob/master/CHANGELOG.md) we keep in the hdf5 crate. For me *personally*, if I'm an active user of some crate, I would be more than happy happy if the devs shared all the new functionality and features that may be relevant to me as a user. As long as it's written in a structured form in human language and doesn't look like [Arrow changelog](https://raw.githubusercontent.com/apache/arrow/master/CHANGELOG.md).

In cases like this, I suggest to always commit as much information as possible to the changelog, and before the release to shorten it down if needed and prettify it (e.g. "10 new methods added to `PyArrayDescr` to match `np.dtype` API, see docs for details"). It's usually easier to do it this way than the other way around (hunting through other contributors' commits in the commit history to see if there's anything worth noting in the changelog before the release).

Looking at the current `rust-numpy`'s changelog, it looks like there has been near zero development from 0.12 to 0.15 (all notes are brief 'fixed bug', 'version bump'). And yet there's around a hundred commits and a substantial a lot of work. In fact, I'd even go as far as update the previous entries in the changelog to expand them slightly and make the changelog more self-explanatory (e.g. instead of 'added einsum' -> 'added einsum!() macro equivalent to calling np.einsum()`, etc).

But anyways, I'm more of a newcomer here so that's my personal 2 cents on the subject :)
